crypto/thread/internal.c

```diff
@@ -0,0 +1,253 @@
+{
+    CRYPTO_TASK *task;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    task = (CRYPTO_TASK *)vtask;
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL || vtask == NULL)
+        return NULL;
+
+    CRYPTO_MUTEX_lock(task->lock);
+    {
+        CRYPTO_MUTEX_lock(tdata->lock);
+        {
+            task->id = tdata->tasks.created_tasks++;
+            if (thread_can_spawn(tdata) == 0)
+                thread_schedule(tdata, task);
+            else if (thread_start_now(ctx, task) == 0)
+                thread_schedule(tdata, task);
+        }
+        CRYPTO_MUTEX_unlock(tdata->lock);
+
+        while ((options & CRYPTO_THREAD_START_AWAIT) != 0) {
+            CRYPTO_mem_barrier();
+            if (task->state >= CRYPTO_TASK_STARTED)
+                break;
+            CRYPTO_CONDVAR_wait(task->cond, task->lock);
+        }
+    }
+    CRYPTO_MUTEX_unlock(task->lock);
+
+    return (void *) task;
+}
+
+int crypto_thread_join(OSSL_LIB_CTX *ctx, void *vtask,
+                       CRYPTO_THREAD_RETVAL *retval)
+{
+    CRYPTO_TASK *task;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    task = (CRYPTO_TASK *) vtask;
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (task == NULL || tdata == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(task->lock);
+    while (1) {
+        CRYPTO_mem_barrier();
+        if (task->state == CRYPTO_TASK_FINISHED)
+            break;
+        CRYPTO_CONDVAR_wait(task->cond, task->lock);
+    }
+
+    if (retval != NULL)
+        *retval = task->retval;
+
+    CRYPTO_MUTEX_unlock(task->lock);
+
+    return 1;
+}
+
+size_t crypto_thread_get_available_threads(OSSL_LIB_CTX *ctx)
+{
+    size_t retval;
+    OSSL_LIB_CTX_THREADS *t;
+
+    t = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (t == NULL)
+        return 0;
+
+    if (t->enabled == 0)
+        return 0;
+
+    if (t->workers.internal.cap < 0)
+        return -1;
+
+    CRYPTO_MUTEX_lock(t->lock);
+    retval = t->workers.internal.cap - list_size(&t->workers.internal.busy) +
+        list_size(&t->workers.external.available);
+    CRYPTO_MUTEX_unlock(t->lock);
+    return retval;
+}
+
+int thread_spawn_worker_task(OSSL_LIB_CTX *ctx, CRYPTO_WORKER_CALLBACK cb,
+                             void *vtask)
+{
+    CRYPTO_WORKER *worker;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL || tdata->enabled == 0)
+        return 0;
+
+    worker = CRYPTO_WORKER_new(ctx, cb, vtask);
+    if (worker == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(worker->lock);
+    worker->handle = crypto_thread_native_start(worker_main, (void *)worker, 0);
+    if (worker->handle == NULL) {
+        CRYPTO_MUTEX_unlock(worker->lock);
+        CRYPTO_WORKER_free(worker);
+        return 0;
+    }
+
+    while (worker->initialized == 0)
+        CRYPTO_CONDVAR_wait(worker->cond, worker->lock);
+
+    CRYPTO_MUTEX_unlock(worker->lock);
+
+    return 1;
+}
+
```

Comment:
nit : two blank lines, while one is the regular usage in this code.

Reply:
Thanks! Fixed.